universum/modules/output/base_output.py

```diff
@@ -3,52 +3,138 @@
 
 class BaseOutput(Module):
     """
-    Abstract base class for output drivers
+    Abstract base class for output drivers.  Defines methods that have to be implemented by each module that
+    implements a type of output.
     """
 
-    def open_block(self, num_str, name):
+    def log_execution_start(self, title: str, version: str) -> None:
+        """
+        Print the universum startup message, for example:
+        "==> Universum xx.yy.zz started execution"
+
+        :param title: name of the product
+        :param version: version of the product
+        """
+        self.log(self._build_execution_start_msg(title, version))
+
+    def log_execution_finish(self, title: str, version: str) -> None:
+        """
+        Print the universum finished message, for example:
+        "==> Universum xx.yy.zz finished execution"
+
+        :param title: name of the product
+        :param version: version of the product
+        """
+        self.log(self._build_execution_finish_msg(title, version))
+
+    def log(self, line: str) -> None:
+        """
+        Print a line to the output. The most generic type of log.
+        Mostly used for printing own messages of universum.
+
+        :param line: the line to print
+        """
         raise NotImplementedError
 
-    def close_block(self, num_str, name, status):
+    def log_error(self, description: str) -> None:
+        """
+        Print an error message to the output. Error may use different colors or other formatting.
+
+        :param description: the detailed error message
+        """
         raise NotImplementedError
 
-    def report_error(self, description):
+    def log_external_command(self, command: str) -> None:
+        """
+        Print the start of the external command with parameters.
+
+        :param command: external command with parameters
+        """
         raise NotImplementedError
 
-    def report_skipped(self, message):
+    def log_stdout(self, line: str) -> None:
+        """
+        Print one line of the standard output of the previously launched external command.
+
+        :param line: the line to print
+        """
         raise NotImplementedError
 
-    def report_step(self, message, status):
-        self.log(message)
+    def log_stderr(self, line: str) -> None:
```

Comment:
I find that log_stderr and log_error lack distinction, could we make use of each a bit clearer?

Reply:
same with log and log_stdout

Reply:
Isn't the doc string enough?
The idea is that log_stdout and log_stderr used for printing respective output of external commands.
log and log_error are for producing own messages.